cmd/gen-go-sample/inittree.go

```diff
@@ -189,7 +155,66 @@ equal:
 	if sc.Scan() != scanner.EOF {
 		return report(errors.E(nil, "expected EOF, found %q", sc.TokenText()))
 	}
-	return err
+	return report(nil)
+}
+
+func (t *initTree) parseSampleArgPath(txt string, info pbinfo.Info, varName string) (*initTree, error) {
+	sc, report := initScanner("." + txt)
+	t, _, err := t.parsePath(sc, info)
+	if err != nil {
+		return nil, report(err)
+	}
+
+	var cp initTree
+	cp = *t
+	*t = initTree{leafVal: varName}
+
+	return &cp, report(nil)
+}
+
+// parsePath parses a path as defined below. It returns the subtree specified by the path,
+// the last scanned token, and any error.
+//
+// path = ident { '.' ident } .
+func (t *initTree) parsePath(sc *scanner.Scanner, info pbinfo.Info) (*initTree, rune, error) {
+	for {
+		switch r := sc.Scan(); r {
+		case '.':
+			if r := sc.Scan(); r != scanner.Ident {
+				return nil, r, errors.E(nil, "expected ident, found %q", sc.TokenText())
+			}
+			if t2, err := t.get(sc.TokenText(), info); err != nil {
+				return nil, 0, err
+			} else {
+				t = t2
+			}
+
+		default:
+			return t, r, nil
+		}
+	}
+}
+
+func initScanner(s string) (*scanner.Scanner, func(error) error) {
+	var sc scanner.Scanner
+	sc.Init(strings.NewReader(s))
+	sc.Mode = scanner.ScanIdents | scanner.ScanInts | scanner.ScanFloats | scanner.ScanStrings
+
+	// Scanner reports error by calling sc.Error; we save the error so we always report
+	// the first.
+	var err error
+	report := func(e error) error {
```

Comment:
It seems this method returned does not have much to do with the outer scope? Is it OK to just make it a standalone helper function instead of a function returned by `initScanner`?